File: coding/src/com/algos/core/li15_dp/revised/Grid_L329_LongestIncreasingPathInAMatrix.java
package  com.algos.core.li15_dp.revised;

public class Grid_L329_LongestIncreasingPathInAMatrix {

	public static void main(String[] args) {
		Grid_L329_LongestIncreasingPathInAMatrix obj = new Grid_L329_LongestIncreasingPathInAMatrix();

		int[][] matrix = new int[][]{{3,4,5},{3,2,6},{2,2,1}};
		System.out.println(obj.longestIncreasingPath(matrix));
	}

	// Problem: https://leetcode.com/problems/longest-increasing-path-in-a-matrix/
	// Idea: Run dfs from each cell in the direction of increasing path.
	// Store the path lens in a lip matrix so that we don't have to compute dfs if it is already done.
	// Path len for a cell is 1+max of all its neighbors. Find it and store it in lip while recursion exiting.
	public int longestIncreasingPath(int[][] matrix) {
		int globalMax = 0, m = matrix.length, n = matrix[0].length;

		int[][] dp = new int[m][n];
		for (int i = 0; i < matrix.length; i++) {
			for (int j = 0; j < matrix[i].length; j++) {
				int max = dfs(i, j, matrix, dp);
				if(max > globalMax)
					globalMax = max;
			}
		}
		
		return globalMax;
	}

	private int dfs(int i, int j, int[][] matrix, int[][] dp) {
		if(dp[i][j] != 0) 
			return dp[i][j];
		
		int max = 0;
		int[][] dirs = new int[][] {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
		for (int[] dir : dirs) {
			int ni = i + dir[0], nj = j + dir[1];
			if (isSafe(i, j, ni, nj, matrix)) {
				int mx = dfs(ni, nj, matrix, dp);
				max = Math.max(max, mx);
			}
		}
		dp[i][j] = 1 + max;
		return dp[i][j];
	}

	private boolean isSafe(int i, int j, int ni, int nj, int[][] matrix) {
		return ni >= 0 && ni < matrix.length && nj >= 0 && nj < matrix[0].length 
				&& matrix[ni][nj] > matrix[i][j];
	}
}

File: dsa/topics/15_dp/python/grid_l329_longest_increasing_path_in_amatrix.py
# Problem: https://leetcode.com/problems/longest-increasing-path-in-a-matrix/
# Idea: Run dfs from each cell in the direction of increasing path.
# Store the path lens in a lip matrix so that we don't have to compute dfs if it is already done.
# Path len for a cell is 1+max of all its neighbors. Find it and store it in lip while recursion exiting.


def longestIncreasingPath(matrix):
    globalMax = 0
    m = len(matrix)
    n = len(matrix[0])

    dp = [[0] * n for _ in range(m)]
    for i in range(m):
        for j in range(n):
            maxLen = dfs(i, j, matrix, dp)
            if maxLen > globalMax:
                globalMax = maxLen

    return globalMax


def dfs(i, j, matrix, dp):
    if dp[i][j] != 0:
        return dp[i][j]

    maxLen = 0
    dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]
    for di, dj in dirs:
        ni = i + di
        nj = j + dj
        if isSafe(i, j, ni, nj, matrix):
            curr = dfs(ni, nj, matrix, dp)
            maxLen = max(maxLen, curr)

    dp[i][j] = 1 + maxLen
    return dp[i][j]


def isSafe(i, j, ni, nj, matrix):
    return (
        ni >= 0 and ni < len(matrix)
        and nj >= 0 and nj < len(matrix[0])
        and matrix[ni][nj] > matrix[i][j]
    )


# ---------------- Test (same as Java main) ----------------

matrix = [
    [3, 4, 5],
    [3, 2, 6],
    [2, 2, 1]
]

print(longestIncreasingPath(matrix))
